src/NCronJob/Scheduler/ObservablePriorityQueue.cs

```diff
@@ -0,0 +1,143 @@
+    }
+
+    public ObservablePriorityQueue(IComparer<TPriority> comparer, IEnumerable<(TElement, TPriority)> items)
+    {
+        ArgumentNullException.ThrowIfNull(comparer);
+        PriorityQueue = new PriorityQueue<TElement, TPriority>(comparer);
+
+        foreach (var (element, priority) in items)
+        {
+            Enqueue(element!, priority);
+        }
+    }
+
+    public event NotifyCollectionChangedEventHandler? CollectionChanged;
+
+    public void Enqueue([DisallowNull] TElement element, [DisallowNull] TPriority priority)
+    {
+        ArgumentNullException.ThrowIfNull(element);
+        ArgumentNullException.ThrowIfNull(priority);
+
+        lock (Lock)
+        {
+            PriorityQueue.Enqueue(element, priority);
+            OnCollectionChanged(new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Add, element));
+        }
+    }
+
+    public TElement Dequeue()
+    {
+        lock (Lock)
+        {
+            if (PriorityQueue.Count == 0) throw new InvalidOperationException("Queue is empty");
+
+            var element = PriorityQueue.Dequeue();
+            OnCollectionChanged(new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Remove, element));
+
+            return element;
+        }
+    }
+
+    public bool TryDequeue([MaybeNullWhen(false)] out TElement element, [MaybeNullWhen(false)] out TPriority priority)
+    {
+        lock (Lock)
+        {
+            if (PriorityQueue.Count == 0)
+            {
+                element = default;
+                priority = default;
+                return false;
+            }
+
+            if(PriorityQueue.TryDequeue(out element,out priority))
+            {
+                OnCollectionChanged(new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Remove, element));
+                return true;
+            }
+
+            return false;
+        }
+    }
+
+    public bool TryPeek([MaybeNullWhen(false)] out TElement element, [MaybeNullWhen(false)] out TPriority priority)
+    {
+        lock (Lock)
+        {
+            if (PriorityQueue.Count == 0)
+            {
+                element = default;
+                priority = default;
+                return false;
+            }
+
+            return PriorityQueue.TryPeek(out element, out priority);
+        }
+    }
+
+    public void Clear()
+    {
+        lock (Lock)
+        {
+            PriorityQueue.Clear();
+            OnCollectionChanged(new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Reset));
+        }
+    }
+
+    public int Count
+    {
+        get
+        {
+            lock (Lock)
+            {
+                return PriorityQueue.Count;
+            }
+        }
+    }
+
+    private void OnCollectionChanged(NotifyCollectionChangedEventArgs args)
+    {
+        var handler = Interlocked.CompareExchange(ref CollectionChanged, null, null);
```

Comment:
Is this to make the function thread-safe?

Reply:
Yes exactly, it retrieves the current value of CollectionChanged in a thread-safe manner, making sure that the value isn't changed by another thread between the time it's checked and the time it's used.

Reply:
Thanks for clearing that up